packages/pg-cursor/index.js

```diff
@@ -6,231 +6,232 @@ const util = require('util')
-}
+    this.state = 'error'
+    this._error = msg
+    // satisfy any waiting callback
+    if (this._cb) {
+      this._cb(msg)
+    }
+    // dispatch error to all waiting callbacks
+    for (let i = 0; i < this._queue.length; i++) {
+      this._queue.pop()[1](msg)
+    }
 
-Cursor.prototype.handleError = function (msg) {
-  // If we're in an initialized state we've never been submitted
-  // and don't have a connection instance reference yet.
-  // This can happen if you queue a stream and close the client before
-  // the client has submitted the stream.  In this scenario we don't have
-  // a connection so there's nothing to unsubscribe from.
-  if (this.state !== 'initialized') {
-    this.connection.removeListener('noData', this._ifNoData)
-    this.connection.removeListener('rowDescription', this._rowDescription)
-    // call sync to trigger a readyForQuery
-    this.connection.sync()
+    if (this.listenerCount('error') > 0) {
+      // only dispatch error events if we have a listener
+      this.emit('error', msg)
+    }
   }
 
-  this.state = 'error'
-  this._error = msg
-  // satisfy any waiting callback
-  if (this._cb) {
-    this._cb(msg)
-  }
-  // dispatch error to all waiting callbacks
-  for (let i = 0; i < this._queue.length; i++) {
-    this._queue.pop()[1](msg)
+  _getRows(rows, cb) {
+    this.state = 'busy'
+    this._cb = cb
+    this._rows = []
+    const msg = {
+      portal: this._portal,
+      rows: rows,
+    }
+    this.connection.execute(msg, true)
+    this.connection.flush()
   }
 
-  if (this.listenerCount('error') > 0) {
-    // only dispatch error events if we have a listener
-    this.emit('error', msg)
+  close(cb) {
+    if (!this.connection || this.state === 'done') {
+      if (cb) {
+        return setImmediate(cb)
+      } else {
+        return
+      }
+    }
+
+    this._closePortal()
+    this.state = 'done'
+    if (cb) {
+      this.connection.once('readyForQuery', function () {
+        cb()
+      })
+    }
   }
-}
 
-Cursor.prototype._getRows = function (rows, cb) {
-  this.state = 'busy'
-  this._cb = cb
-  this._rows = []
-  const msg = {
-    portal: this._portal,
-    rows: rows,
+  read(rows, cb) {
+    if (this.state === 'idle' || this.state === 'submitted') {
+      return this._getRows(rows, cb)
+    }
+    if (this.state === 'busy' || this.state === 'initialized') {
+      return this._queue.push([rows, cb])
+    }
+    if (this.state === 'error') {
+      return setImmediate(() => cb(this._error))
+    }
+    if (this.state === 'done') {
+      return setImmediate(() => cb(null, []))
+    } else {
+      throw new Error('Unknown state: ' + this.state)
+    }
   }
-  this.connection.execute(msg, true)
-  this.connection.flush()
 }
 
-// users really shouldn't be calling 'end' here and terminating a connection to postgres
-// via the low level connection.end api
+// we can't correctly use util.deprecate in an ES6 method
 Cursor.prototype.end = util.deprecate(function (cb) {
```

Comment:
You can keep defining it inside the class, it just needs this afterwards:

```js
Cursor.prototype.end = util.deprecate(Cursor.prototype.end, 'Cursor.end is deprecated. Call end on the client itself to end a connection to the database.')
```

Reply:
Oh perfect, that's much better!